build/images/katas/Dockerfile

```diff
@@ -0,0 +1,67 @@
+# We use the iqsharp-base image, as that includes
+# the .NET Core SDK, IQ#, and Jupyter Notebook already
+# installed for us.
+FROM mcr.microsoft.com/quantum/iqsharp-base:0.11.2004.2715-beta
+
+# Add metadata indicating that this image is used for the katas.
+ENV IQSHARP_HOSTING_ENV=KATAS_DOCKERFILE
+
+# Make sure the contents of our repo are in ${HOME}
+# Required for mybinder.org
+COPY . ${HOME}
+
+USER root
+RUN chown -R ${USER} ${HOME} && \
+    chmod +x ${HOME}/scripts/*.sh
+
+# Install Python dependencies for the Python visualization and tutorial notebooks
+RUN pip install -I \
+        matplotlib \
+        pytest
+
+USER ${USER}
+
+# Pre-exec notebooks to improve first-use start time
+RUN cd ${HOME} && \
+    ./scripts/prebuild-kata.sh BasicGates && \
```

Comment:
Looking at the bigger picture of this PR and #335, what happens (once these two changes are in) when we update one of the existing katas or notebooks or add a new one (such as the QFT kata right now)? Are we planning to build a "released" version of the katas with the release which is going to be the base, and then any changes after that are rebuilt by Binder when a specific kata is executed? Will the new kata have access to Nuget packages downloaded by prebuilding other katas?

Reply:
Great question that requires a longer answer.

**About the image**

With a Katas image, it will only contain what was available when that image was built, so we would have a monthly Katas release aligned with QDK.
Whatever changes happened between when the image was built and the next release won't be available in MyBinder, as it would normally reuse the pre-built image the the "official" release.

Some options to consider:
1) We could also have another branch where MyBinder will pick the latest changes in master (even before the monthly QDK release), but that would require some automation to keep that branch in sync with master.
2) Alternatively, we could have #335 in a special branch that is used by MyBinder to fetch the pre-built image instead of it being the default mode. I think this is what @cgranade did for the samples.

Now I think that option 2 is likely the best one as it keeps the current Katas behavior but offers an option for an optimized MyBinder experience.

**About the Nuget packages**

About the Nuget packages, the pre-built image will pre-download all packages referenced by all Katas at the moment of the build, and then it disable online packages (so if users try to load packages that were not originally referenced by any Katas, that package won't be loaded). 

Alternatively, we could allow the pre-built image to fetching online packages, and only remove that option for the optimized MyBinder experience with #335. 

**Conclusion**

With that, I think it's best to leave this original Dockerfile in the root as is (so people can still attempt to run it on MyBinder) and offer an optimized Dockerfile in a separate KatasImage branch that will point to the pre-built image based on the original root Dockerfile.